design-documents/frontend/advanced-js-bundling/advanced-js-bundling.md

```diff
@@ -0,0 +1,82 @@
+
+Better JS bundling approach can reduce package size by several times. This is one of the most valuable enhancement can be done from storefront side performance perspective. According to our measurements, advanced bundling can improve overall performance score, main-thread work time and TTI up to 2-3x time.
+
+## Solution
+
+I propose to introduce 'Bundlegento' as separate tool for parsing Magento store pages and generate page specific bundles.
+Currently we have prototype written in NodeJS that parse HTML pages and can generate needed bundles for Magento store. Since that this is not version specific solution, it will support all Magento released versions
+
+Also we will need to add needed functionality inside Magento that will load needed JS bundle according to layout handlers was used. JS bundle files will be added under git track and stored in theme specific location
+
+## Usage
+
+'Bundlegento' will generate needed bundles according to its configuration, then JS files will be added to theme directory and under git control manually. On static generation process it will be transferred to `pub/static/<theme>` files and used on Magento store-front
+
+### Fetching
+
+Because static analysis cannot be relied on, `bundlegento` eschews it entirely and instead uses runtime module data it collects from different pages of a store. The pages that are crawled are specified by a user of `bundlegento` in their [configuration file](CONFIG.md).
+
+[`Puppeteer`](https://pptr.dev/) is used to automate a headless version of [Google Chrome](https://www.google.com/chrome/). As each URL from the config is visited, JavaScript snippet is injected into the page that implements the [`onResourceLoad`](https://github.com/requirejs/requirejs/wiki/internal-api:-onresourceload) handler of `RequireJS`. This code runs before any other scripts loaded by the store, and ensures that every loaded module is accounted for.
+
+Internally, modules are stored per-group, instead of per-URL. This is because a single page does not get its own bundle - a _group_ gets a bundle.
+
+Note that the tool does _not_ keep an actual graph of modules, but instead maintains a flat list. The graph is not necessary because we have an entire list of modules used on a page, allowing us to skip any graph traversal.
+
+### Splitting
+
+With module usage data obtained during the fetching phase, modules can now be compared between groups to determine which are unique to a page, and which should be broken out to shared bundles.
+
+The algorithm for splitting modules is fairly basic:
+
+-   If a module exists in exactly 1 group, keep it in that group's bundle
+-   If a module exists in exactly 2 groups, move it to a shared file that is loaded for any page in those 2 groups
+-   If a module exists in >= 3 groups, move it to a shared bundle that will be loaded on all pages
```

Comment:
Taken that I can totally understand the reasoning behind common bundle, getting the grasp of shared ones is difficult.

Would it be true that if we have 5 groups we would end up with 10 shared bundles, 15 for 6, 21 for 7 etc.?

Is so, then I'm afraid it would negatively impact the performance across the pages as opposed to having one common bundle loaded from cache for every other group and one extra file loaded specific to each of them.